shared/utility/math/filter/ParticleFilter.h

```diff
@@ -60,20 +60,40 @@ namespace math {
                 reset(initialMean, initialCovariance, number_of_particles_);
             }
 
-            void reset(StateVec initialMean, StateMat initialCovariance, int number_of_particles_) {
-                particles = arma::zeros(number_of_particles_, Model::size);
-                setState(initialMean, initialCovariance);
+            void reset(StateVec initialMean, StateMat initialCovariance, int number_of_particles_ = 100) {
+                particles = getParticles(initialMean, initialCovariance, number_of_particles_);
             }
 
-            void setState(StateVec initialMean, StateMat initialCovariance) {
+            void resetAmbiguous(std::vector<StateVec> initialMean,
+                                std::vector<StateMat> initialCovariance,
+                                int number_of_particles_ = 100) {
+                if (initialMean.size() != initialCovariance.size()) {
+                    throw std::runtime_error(std::string(__FILE__) + " : " + std::to_string(__LINE__)
+                                             + " different number of means vs covariances provided.");
+                }
+                // Sample the same number of particles for each possibility
+                particles =
```

Comment:
This and other consecutive joins would be much faster and cleaner as a lambda (if this doesn't compile Trent will know how to make it work):
particles = arma::mat(number_of_particles_ / initialMean.size(), initialMean.size());
int i = 0;
particles.each_col([&i](arma::col& result) {
result =  getParticles(initialMean[i], initialCovariance[i], number_of_particles_ / initialMean.size());
++i;
});